File: 3997-count-dominant-nodes-in-a-binary-tree/3997-count-dominant-nodes-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans=0;
    vector<TreeNode*> allNodes;
    void inorder(TreeNode* root){
        if(root==nullptr)
            return;
        if(root->left!=nullptr)
            inorder(root->left);
        allNodes.push_back(root);
        if(root->right!=nullptr)
            inorder(root->right);
    }

    int maxi(TreeNode *root,int maxVal){
        if(root==nullptr)
            return maxVal;

        if(root->left!=nullptr){
            maxVal = maxi(root->left,max(root->val,maxVal));
        }
            maxVal = max(maxVal,root->val);

        if(root->right!=nullptr)
            maxVal = maxi(root->right,max(root->val,maxVal));
        
        return maxVal;
    }

    int countDominantNodes(TreeNode* root) {
        inorder(root);
        for(int i=0;i<allNodes.size();i++){
            int val = allNodes[i]->val;
            int maxVal = maxi(allNodes[i],val);
            if(val==maxVal)
                ans++;
        }
        return ans;
    }
};

File: 3997-count-dominant-nodes-in-a-binary-tree/3997-count-dominant-nodes-in-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def __init__(self):
        self.ans = 0
        self.allNodes = []

    def inorder(self, root: TreeNode) -> None:
        if root is None:
            return
        if root.left is not None:
            self.inorder(root.left)
        self.allNodes.append(root)
        if root.right is not None:
            self.inorder(root.right)

    def maxi(self, root: TreeNode, maxVal: int) -> int:
        if root is None:
            return maxVal

        if root.left is not None:
            maxVal = self.maxi(root.left, max(root.val, maxVal))
        
        maxVal = max(maxVal, root.val)

        if root.right is not None:
            maxVal = self.maxi(root.right, max(root.val, maxVal))
        
        return maxVal

    def countDominantNodes(self, root: TreeNode) -> int:
        self.inorder(root)
        for i in range(len(self.allNodes)):
            val = self.allNodes[i].val
            maxVal = self.maxi(self.allNodes[i], val)
            if val == maxVal:
                self.ans += 1
        return self.ans